### ISSUE TITLE
NNBD: Map [] operator accepts null even when keys are non-nullable

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#3049]
Consider the following:

```dart
void main() {
  var map = <String, String>{};

  map[null];
}
```

The current behavior is, `map[null]` compiles (and return `null`)

I would expect it to emit a compilation error instead.

----

This appears to be voluntary, considering `ma[null] = ` correctly does not compile.

Maybe we should have a lint?

**Comment by srawlins** [comment]
This is implemented in [collection_methods_unrelated_type](https://dart-lang.github.io/linter/lints/collection_methods_unrelated_type.html).
